src/groupeddataframe/splitapplycombine.jl

```diff
@@ -962,14 +1056,51 @@ function _combine(f::AbstractVector{<:Pair},
                 throw(ArgumentError("a single value or vector result is required when passing " *
                                     "multiple functions (got $(typeof(res)))"))
             end
+            if !(firstres isa AbstractVector) && isnothing(idx_agg)
+                idx_agg = Vector{Int}(undef, length(gd))
+                fillfirst!(nothing, idx_agg, 1:length(gd.groups), gd)
+            end
             idx, outcols, _ = _combine_with_first(wrap(firstres), fun, gd, incols,
-                                                  Val(firstmulticol))
+                                                  Val(firstmulticol),
+                                                  firstres isa AbstractVector ? nothing : idx_agg)
+            @assert length(outcols) == 1
             res[i] = idx, outcols[1]
         end
     end
-    # TODO: avoid recomputing idx for each pair when !all(isagg, f)
```

Comment:
This one is still valid I think (the redundant computation happens in `_combine_with_first`, so you could move it there).

Reply:
1. if functions return a single row we compute `idx_agg` only once
2. for all functions returning vectors each of them potentially can return a different `idx` and only later we can find if they have the same lengths or not.

What could be optimized is that for functions returning vectors we could work like this:
1. for a first function like this compute `idx` as is now in `_combine_tables_with_first`
2. for any other function have a different `_combine_tables_with_first` implementation that takes the `idx` vector computed in step 1 and instead of creating it only check if it would be consistent with what `idx` holds.

I will add such a comment but leave it as a future optimization.

Reply:
Just to be clear - we currently avoid multiple allocation of `idx` if single-row columns are given (and this is then `idx_agg`). I kept `idx_agg` name as single-row columns are essentially aggregations of groups.

I leave it unresolved as again - this is a crucial concept of this PR.